juju/juju
# Verify Endpoint Bindings Integrity When Moving Subnets

## Description of change

This patch adds a verification to ensure that the integrity of endpoint bindings is maintained when moving subnets between spaces.

This is the first of two patches for the functionality and simply considers violation to be when not all units of the application are connected to a bound space.

Subsequent logic will interrogate the unit connectivity, and if the binding can be preserved by changing it to the destination space, that action will be taken.

## QA steps

- Bootstrap to AWS
- Carve one of the subnets into a new space. E.g. `juju add-space beta 172.31.16.0/20`.
- `juju deploy postgresql --bind "beta" --to subnet=172.31.16.0/20`.
- `juju move-to-space alpha 172.31.16.0/20`.
- Error should read:
```bash
ERROR cannot update space "alpha": moving subnet(s) to space "alpha" violates endpoint binding db:beta for application "postgresql"
        units not connected to the space: postgresql/0
```
- Other subnets can be moved; these should both succeed:
  - `juju move-to-space beta 172.31.0.0/20`.
  - `juju move-to-space alpha 172.31.0.0/20`.

## Documentation changes

None.

## Bug reference

N/A
